lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1519,7 +1515,84 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
       // capturing it will create a reference cycle.
       if (!ty->hasReferenceSemantics())
         return false;
-
+      
+      // If this is an implicit self parameter to a `AbstractFunctionDecl`
+      // (`FuncDecl`, `ConstructorDecl`, or `DestructorDecl`,
+      //  `@autoclosure @escaping () -> String = String()` as one example)
+      // then it isn't actually capturing the closure's 'self', and is fine.
+      if (isa<AbstractFunctionDecl>(DRE->getDecl())) {
+        return false;
+      }
+      
+      // If this decl isn't named "self", then it isn't an implicit self capture
+      // and we have no reason to reject it.
+      if (!DRE->getDecl()->getBaseName().getIdentifier().is("self")) {
```

Comment:
Does `VarDecl::isSelfParamCapture` work here?

Reply:
I checked and changing this to use `VarDecl::isSelfParamCapture` causes a bunch of test failures that currently pass. 

That may be because this is currently checking a `ValueDecl`, which has to be casted to a `VarDecl` for the `isSelfParamCapture` check, causing the check to be skipped in cases when we don't have a `VarDecl`.